### PR TITLE
[API Compatibility] Add `paddle.compat.nn.functional.linear`

### PR DESCRIPTION
**Pull request by Enigmatisms**
### PR Category
Operator Mechanism

### PR Types
New features

### Description
Torch 兼容的 `paddle.compat.nn.functional.linear`。实现是 _C_ops.matmul (transpose_y = True) + _C_ops.add。PaConvert 已通过。性能数据（与torch对比）：

```
统计信息:
平均前向加速比: 1.10x
平均反向加速比: 1.11x
平均总加速比: 1.08x
最佳性能案例: 极小尺寸-有偏置 (加速比: 1.39x)
最差性能案例: 高维输入-有偏置 (加速比: 0.61x)

偏置对性能影响:
有偏置案例平均加速比: 0.97x
无偏置案例平均加速比: 1.19x
偏置带来的性能差异: -0.22x

维度对性能影响:
2D输入平均加速比: 1.14x
3D输入平均加速比: 0.99x
4D输入平均加速比: 1.01x
5D输入平均加速比: 1.18x
```

此外，调整了：
- https://github.com/PaddlePaddle/Paddle/pull/76121

引入的 `paddle.compat.softmax`。`paddle.softmax` 已经存在，并且基于 `compat_softmax`，所以 `paddle.compat.softmax` 没有存在的必要。还调整了一下几个 `__init__.py` 的 import 格式。

Pcard-89620

### CODE REVIEW

File: python/paddle/compat/nn/functional/__init__.py [hunk 1]
```diff
@@ -38,7 +38,7 @@
 
 from paddle.utils.decorator_utils import ForbidKeywordsDecorator
 
-__all__ = ['pad', 'softmax']
+__all__ = ['pad', 'softmax', 'linear']
```
Comment (zhwesky2010) [inline python/paddle/compat/nn/functional/__init__.py:41]: 后面移除掉 paddle/compat/\___init\___.py 中的softmax吧，没有`paddle.compat.softmax`，只有`paddle.compat.nn.functional.softmax`

File: python/paddle/nn/functional/common.py [hunk 2]
```diff
@@ -2359,6 +2361,12 @@ def cosine_similarity(
     return cos_sim
 
 
+@ForbidKeywordsDecorator(
```
Comment (zhwesky2010) [inline python/paddle/nn/functional/common.py]: 这个是给`paddle.nn.Linear`用的，应该只给`paddle.nn.Linear`打1次warning就行。

这两套预计会长期并存，`paddle.nn.Linear` or `paddle.compat.nn.Linear`，不打warning是不是要好些。

File: python/paddle/nn/functional/common.py [hunk 3]
```diff
@@ -74,6 +74,8 @@
         'valid', 'same', 'VALID', 'SAME'
     ]
 
+from paddle.utils.decorator_utils import ForbidKeywordsDecorator
```
Comment (SigureMo) [inline python/paddle/nn/functional/common.py]: import 放在 `TYPE_CHECKING` block 之上

File: python/paddle/nn/functional/common.py (hunk 2 above)
Comment (Enigmatisms) [reply python/paddle/nn/functional/common.py]: 已删除。有道理，但是如果直接用 F.linear 还是有混用风险。装饰器本身带来的overhead是比较小的。

File: python/paddle/compat/nn/functional/__init__.py (hunk 1 above)
Comment (Enigmatisms) [reply python/paddle/compat/nn/functional/__init__.py:41]: 已经在本PR中删除。

File: python/paddle/nn/functional/common.py (hunk 3 above)
Comment (Enigmatisms) [reply python/paddle/nn/functional/common.py]: done

**Review by SigureMo — DISMISSED** [review]

**Review by zhwesky2010 — APPROVED** [review]
LGTM

**Review by SigureMo — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM
